File: internal/domain/exercise.go
package domain

import (
	"strings"
	"time"
)

// --- Value Objects ---

// TrackingType determines which fields are relevant for an exercise.
type TrackingType string

const (
	TrackingWeightReps     TrackingType = "weight_reps"
	TrackingBodyweightReps TrackingType = "bodyweight_reps"
	TrackingDuration       TrackingType = "duration"
	TrackingDistance        TrackingType = "distance"
)

// IsValid returns true if t is a recognized tracking type.
func (t TrackingType) IsValid() bool {
	switch t {
	case TrackingWeightReps, TrackingBodyweightReps, TrackingDuration, TrackingDistance:
		return true
	}
	return false
}

// ValidMuscleGroups is the exhaustive list of allowed muscle group values.
var ValidMuscleGroups = []string{
	"chest", "back", "legs", "shoulders",
	"biceps", "triceps", "core", "cardio", "other",
}

// IsValidMuscleGroup checks whether s is an allowed muscle group value.
func IsValidMuscleGroup(s string) bool {
	for _, v := range ValidMuscleGroups {
		if v == s {
			return true
		}
	}
	return false
}

// ValidEquipment is the exhaustive list of allowed equipment values.
var ValidEquipment = []string{
	"barbell", "dumbbell", "cable", "machine",
	"bodyweight", "band", "kettlebell", "other",
}

// IsValidEquipment checks whether s is an allowed equipment value.
func IsValidEquipment(s string) bool {
	for _, v := range ValidEquipment {
		if v == s {
			return true
		}
	}
	return false
}

// --- Entity ---

// Exercise is a named movement that can be added to program sections
// and tracked in sessions.
type Exercise struct {
	ID           int64
	UUID         string
	Name         string
	MuscleGroup  *string
	Equipment    *string
	TrackingType TrackingType
	Notes        *string
	IsCustom     bool
	CreatedAt    time.Time
	UpdatedAt    time.Time
	DeletedAt    *time.Time
}

// Validate enforces domain business rules on an Exercise.
func (e *Exercise) Validate() error {
	if strings.TrimSpace(e.Name) == "" {
		return NewValidationError("name", "must not be empty")
	}
	if !e.TrackingType.IsValid() {
		return NewValidationError("tracking_type",
			"must be one of: weight_reps, bodyweight_reps, duration, distance")
	}
	if e.MuscleGroup != nil && !IsValidMuscleGroup(*e.MuscleGroup) {
		return NewValidationError("muscle_group",
			"must be one of: chest, back, legs, shoulders, biceps, triceps, core, cardio, other")
	}
	if e.Equipment != nil && !IsValidEquipment(*e.Equipment) {
		return NewValidationError("equipment",
			"must be one of: barbell, dumbbell, cable, machine, bodyweight, band, kettlebell, other")
	}
	return nil
}

File: app/domain/exercise.ts
// Exercise domain model — mirrors internal/domain/exercise.go.
// Uses interfaces (not classes) for React state compatibility.

import { ValidationError } from "./errors";

// ---------------------------------------------------------------------------
// Value objects
// ---------------------------------------------------------------------------

export type TrackingType =
  | "weight_reps"
  | "bodyweight_reps"
  | "duration"
  | "distance";

export const VALID_TRACKING_TYPES: TrackingType[] = [
  "weight_reps",
  "bodyweight_reps",
  "duration",
  "distance",
];

export const VALID_MUSCLE_GROUPS = [
  "chest",
  "back",
  "legs",
  "shoulders",
  "biceps",
  "triceps",
  "core",
  "cardio",
  "other",
] as const;

export type MuscleGroup = (typeof VALID_MUSCLE_GROUPS)[number];

export const VALID_EQUIPMENT = [
  "barbell",
  "dumbbell",
  "cable",
  "machine",
  "bodyweight",
  "band",
  "kettlebell",
  "other",
] as const;

export type Equipment = (typeof VALID_EQUIPMENT)[number];

// ---------------------------------------------------------------------------
// Entity
// ---------------------------------------------------------------------------

export interface Exercise {
  id: number;
  uuid: string;
  name: string;
  muscle_group: string | null;
  equipment: string | null;
  tracking_type: TrackingType;
  notes: string | null;
  is_custom: boolean;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export function validateExercise(e: {
  name: string;
  tracking_type: string;
  muscle_group?: string | null;
  equipment?: string | null;
}): ValidationError | null {
  if (!e.name || e.name.trim().length === 0) {
    return new ValidationError("name", "name is required");
  }

  if (
    !VALID_TRACKING_TYPES.includes(e.tracking_type as TrackingType)
  ) {
    return new ValidationError(
      "tracking_type",
      `invalid tracking type: ${e.tracking_type}`
    );
  }

  if (
    e.muscle_group &&
    !VALID_MUSCLE_GROUPS.includes(e.muscle_group as MuscleGroup)
  ) {
    return new ValidationError(
      "muscle_group",
      `invalid muscle group: ${e.muscle_group}`
    );
  }

  if (
    e.equipment &&
    !VALID_EQUIPMENT.includes(e.equipment as Equipment)
  ) {
    return new ValidationError(
      "equipment",
      `invalid equipment: ${e.equipment}`
    );
  }

  return null;
}
